### PR TITLE
Pin `chunk` and `index` format to `schema` version.

### PR DESCRIPTION
**Pull request by kavirajk**
We pin all three `Chunk`, `HeadBlock` and `TSDB` Version to `schema` version in period config.

This is the following mapping (after being discussed with @owen-d and @sandeepsukhani )

`v12` (current existing schema) - ChunkFormatV3 (UnorderedHeadBlock) + TSDBv2
`v13` (introducing new schema) - ChunkFormatV4 (UnorderedWithNonIndexedLabelsHeadBlockFmt) + TSDBv3

Note the new schema `v13` supports the latest chunk and index format.

**NOTES for Reviewer**

1. General approach is we removed the idea of `index.LiveFormat`, `chunkenc.DefaultChunkFormat` and `chunkenc.DefaultHeadBlockFmt` and made following two changes. These variables were used before to tie chunk and tsdb formats specific to Loki versions. This PR remove that coupling and pin these formats to `schema` version instead.

    1. These variables were replaced with explicit chunk and index formats within those packages (and it's tests)
    2. If these variables were used outside it's own packages say by ingester, compactor, etc. Then we extract correct chunk and index versions from the `schema` config.

2. Add two methods to `periodConfig`. (1) `ChunkFormat()` returning chunk and head format tied to schema (2) `TSDBFormat()` returning tsdb format tied to schema.

2. Other ideas I thought of doing but didn't end up doing is make `ChunkFormat` and `IndexFormat` as separate type (rather than `byte` and `int` currently. Similar to `HeadBlockFmt` type). But didnt' do it eventually to keep the PR small and don't want to complicate with lots of changes.

4. Moved couple of test cases from `chunkenc` to `config` package, because the test case was actually testing methods on `schemaconfig` and it was creating cycling dependencies.

### CODE REVIEW

**Comment by JStickler** [comment]
Are there any updates we need to make to the documentation when this change goes through?
We document the chunk format in https://grafana.com/docs/loki/next/get-started/architecture/.
1 - Is the current chunk format correct (IIRC there was a recent update to this section of the docs)
2 - Do we need to add the schema version number to the docs?

**Comment by kavirajk** [comment]
>  Is the current chunk format correct (IIRC there was a recent update to this section of the docs)

I think you mean the non-index labels changes went into ChunkV4. I think yes. @sandeepsukhani @vlad-diachenko @salvacorts worked on this.

> Do we need to add the schema version number to the docs?

Yes. Schema v13 needs to be added. I will do that either in this PR or follow up PR.

File: pkg/chunkenc/memchunk.go
Comment (chaudum) [inline pkg/chunkenc/memchunk.go:86]: Can be removed

File: pkg/ingester/checkpoint.go
Comment (chaudum) [inline pkg/ingester/checkpoint.go:106]: Has this the potential to break WAL recovery when the WAL spans across multiple periods with different schemas?

File: pkg/storage/config/schema_config.go
Comment (chaudum) [inline pkg/storage/config/schema_config.go:391]: This logic partially duplicates `pkg/chunkenc/memchunk.go:416-425`

I think it would be great to have a single source of truth for the binding of chunk format to head block format.

File: pkg/storage/stores/indexshipper/compactor/retention/retention_test.go
Comment (chaudum) [inline pkg/storage/stores/indexshipper/compactor/retention/retention_test.go:526]: Should this commented out test already pass? I thought there are no "default" chunk and head formats any more, since they are "encoded" into the schema version.

File: pkg/storage/stores/tsdb/index/index.go
Comment (chaudum) [inline pkg/storage/stores/tsdb/index/index.go:66]: Could we keep the "live" format for tests? That way, we don't need to update index tests every time a new version is introduced.

**Review by salvacorts — COMMENTED** [review]
This is looking great! Thank you Kavi, this will ease our lives a lot when implementing and rolling out new chunk formats 🚀 

I left out some minor comments.

File: pkg/chunkenc/memchunk_test.go
Comment (chaudum) [reply pkg/chunkenc/memchunk_test.go:80]: I like the idea of running tests in all formats.

File: pkg/storage/stores/tsdb/index/index.go
Comment (kavirajk) [reply pkg/storage/stores/tsdb/index/index.go:66]: May be calling it as `latestFormat` instead (private to the package and not exported) may be better I think.

File: pkg/ingester/checkpoint.go
Comment (kavirajk) [reply pkg/ingester/checkpoint.go:106]: Good point. ideally I want to figure out this format from schema config. Will fix it.

File: pkg/storage/stores/tsdb/index/index.go
Comment (kavirajk) [reply pkg/storage/stores/tsdb/index/index.go:66]: Hmm. while implementing it, realized even private `latestFormat` is not needed as all the tests were written to cover both `FormatV2` and `FormatV3` already.

Didn't see the need for it. So removed again. I guess we can add when the needs shows up in future.

File: pkg/storage/stores/indexshipper/compactor/retention/retention_test.go
Comment (kavirajk) [reply pkg/storage/stores/indexshipper/compactor/retention/retention_test.go]: Done.

File: pkg/storage/config/schema_config.go
Comment (kavirajk) [reply pkg/storage/config/schema_config.go]: Good idea. Done.

File: pkg/ingester/checkpoint.go
Comment (kavirajk) [reply pkg/ingester/checkpoint.go]: Fixed it.

File: pkg/chunkenc/memchunk.go
Comment (kavirajk) [reply pkg/chunkenc/memchunk.go]: Done.

File: pkg/storage/stores/indexshipper/compactor/retention/retention_test.go
Comment (kavirajk) [reply pkg/storage/stores/indexshipper/compactor/retention/retention_test.go]: Done.

File: pkg/ingester/stream.go
Comment (kavirajk) [reply pkg/ingester/stream.go:614]: I like the idea. Only problem is once we export such variables outside the `chunkenc`, people use it instead of taking it from `schema` config.

Besides, particularly here this unordered head fmt should also depends on either `chunkv3 or chunkv4` (cannot use normal UnorderedBlockFmt with chunk4 for example). 

I handled it differently by exporting `ChunkHeadFormatFor(chunkfmt)` api from `chunkenc` package.

**Review by sandeepsukhani — COMMENTED** [review]
Nice work Kavi! The changes look good to me, modulo Salva's suggestions which are yet to be resolved. While reviewing the code, I had similar concerns about having to update the tests outside of the `chunkenc` package to use the latest chunk format. I think we can put a `LatestPeriodConfig` or `RecommendedPeriodConfig` somewhere for the tests and use it for tests building chunks. The `chunkenc` package can then have most tests running against all supported chunk formats.

File: pkg/chunkenc/memchunk_test.go
Comment (kavirajk) [reply pkg/chunkenc/memchunk_test.go]: Done.

File: pkg/storage/batch_test.go
Comment (kavirajk) [reply pkg/storage/batch_test.go]: Agree. Added recent set of schemas to test cases.

File: pkg/storage/config/schema_config.go
Comment (kavirajk) [reply pkg/storage/config/schema_config.go]: Done. Make sense.

File: pkg/storage/config/schema_config.go
Comment (kavirajk) [reply pkg/storage/config/schema_config.go]: Done.

File: pkg/storage/hack/main.go
Comment (kavirajk) [reply pkg/storage/hack/main.go]: I explained the problem of exposing such function before. Same applies here. I handled it differently (taking it from schema config) 👍

File: pkg/storage/lazy_chunk_test.go
Comment (kavirajk) [reply pkg/storage/lazy_chunk_test.go]: Done.

File: pkg/storage/stores/tsdb/head_manager_test.go
Comment (kavirajk) [reply pkg/storage/stores/tsdb/head_manager_test.go]: Done.

**Comment by kavirajk** [comment]
> I think we can put a LatestPeriodConfig or RecommendedPeriodConfig somewhere for the tests and use it for tests building chunks. The chunkenc package can then have most tests running against all supported chunk formats.

Thanks for the review @sandeepsukhani. Agree.

Basically I tweaked tests inside `chunkenc` to run tests against all supported chunk formats and outside the `chunkenc` I tweaked tests to run against all recent schemas (v11, v12 and v13)

File: pkg/ingester/recovery.go
Comment (sandeepsukhani) [inline pkg/ingester/recovery.go:213]: this code actually doesn't make sense to me because `isAllowed` here is always going to be false. Anyways, not going to block the PR on this.

Comment (sandeepsukhani) [inline pkg/ingester/recovery.go:213]: this code actually doesn't make sense to me because `isAllowed` here is always going to be false. Anyways, not going to block the PR on this.

**Review by sandeepsukhani — APPROVED** [review]
LGTM

**Comment by kavirajk** [comment]
Will merge this once getting some reviews from @owen-d as well.

File: pkg/ingester/instance.go
Comment (owen-d) [inline pkg/ingester/instance.go:116]: This should be detected when writing a chunk within the `stream`, not on the instance. The `instance` may be exist across schema boundaries. This leaves us with two choices:
1) set the format based on the min timestamp in a chunk when it's cut
2) set the format based on the max timestamp in a chunk when it's cut

I think (1) is the more conservative approach and thus more appropriate. We may write an older format to a newer schema bucket, but we won't write a newer format to an older bucket. To be exact, the chunk is written once then added to the index buckets for different days. I'd rather us have some v13-schema indices reference v12 compatible chunks than vice versa.

**Review by owen-d — COMMENTED** [review]
One fix needed then lgtm

File: pkg/ingester/instance.go
Comment (kavirajk) [reply pkg/ingester/instance.go]: Nice catch. thanks @owen-d for the explanation. 

In fact [my initial commit used "per-stream" chunk format selection](https://github.com/grafana/loki/pull/10213/commits/8b6787fcb8c7d75984be131ea31faeec184af5db#diff-5e8ea6dc696f5e7c429726e1d8c59705869c8a64e1793f399e6e6587b137e4bfL296-L302), but was using `model.Now()` instead of `min/max` timestamp. Makes more sense to use correct timestamp here.

Also agree with `min` timestamp.

**Pull request merged**
